src/Sylius/Bundle/CoreBundle/Applicator/CatalogPromotionApplicator.php

```diff
@@ -13,19 +13,39 @@
 
 namespace Sylius\Bundle\CoreBundle\Applicator;
 
+use Sylius\Component\Core\Model\CatalogPromotionInterface;
 use Sylius\Component\Core\Model\ChannelPricingInterface;
 use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionTranslationInterface;
 
 final class CatalogPromotionApplicator implements CatalogPromotionApplicatorInterface
 {
-    public function applyPercentageDiscount(ProductVariantInterface $variant, float $discount): void
-    {
+    public function applyCatalogPromotion(
+        ProductVariantInterface $variant,
+        CatalogPromotionInterface $catalogPromotion
+    ): void {
         /** @var ChannelPricingInterface $channelPricing */
         foreach ($variant->getChannelPricings() as $channelPricing) {
             if ($channelPricing->getOriginalPrice() === null) {
                 $channelPricing->setOriginalPrice($channelPricing->getPrice());
             }
-            $channelPricing->setPrice((int) ($channelPricing->getPrice() - ($channelPricing->getPrice() * $discount)));
+
+            $channelPricing->setPrice((int) ($channelPricing->getPrice() - ($channelPricing->getPrice() * 0.5)));
+
+            $channelPricing->addAppliedPromotion($this->formatAppliedPromotion($catalogPromotion));
         }
     }
+
+    private function formatAppliedPromotion(CatalogPromotionInterface $catalogPromotion): array
+    {
+        /** @var CatalogPromotionTranslationInterface $translation */
+        $translation = $catalogPromotion->getTranslations()->first();
+
+        /** @var string $label */
+        $label = $translation->getLabel();
+        /** @var string $code */
+        $code = $catalogPromotion->getCode();
+
+        return [$code => ['name' => $label]];
```

Comment:
If we are saving only one label in one translation, there will be no possibility to translate this label on variants 🤔

Reply:
Indeed. Probably we should save the label in each locale - it makes this data structure a little bit more complicated, but benefits are undeniable. I would provide is as an improvement in the separate PR, wdyt?